### PR TITLE
Remove old CTC decoder (Fixes #1675)

### LINKED ISSUE #1675: Write Python/NumPy bindings for ctcdecode and use it where we currently use libctc_decoder_with_kenlm.so
- [x] Write ctcdecode bindings
- [x] Switch Python code to it
- [x] Remove old CTC code & KenLM scorer

### PR DESCRIPTION
**Pull request by reuben**
- [x] Remove old decoder code and build system integration
- [x] Remove references in code and tests
- [x] Update documentation

### CODE REVIEW

**Comment by reuben** [comment]
@lissyx I'm having trouble getting the ctcdecode package to build on ARM64, getting relocation errors in libm.a when doing the final linking of the Python extension. I see that in native_client/BUILD we link against libstdc++.a, is that a requirement on our ARM platforms? The strange thing is, it's building fine on ARMv7...

**Comment by lissyx** [comment]
> @lissyx I'm having trouble getting the ctcdecode package to build on ARM64, getting relocation errors in libm.a when doing the final linking of the Python extension. I see that in native_client/BUILD we link against libstdc++.a, is that a requirement on our ARM platforms? The strange thing is, it's building fine on ARMv7...

They should share the same linking strategy, but maybe you lack some `-PIE` or something like that?

File: taskcluster/test-training_upstream-linux-amd64-py27mu-opt.yml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 build:
   template_file: test-linux-opt-base.tyml
   dependencies:
-    - "linux-amd64-ctc-opt"
+    - "linux-amd64-cpu-opt"
```
Comment (lissyx) [inline taskcluster/test-training_upstream-linux-amd64-py27mu-opt.yml]: I don't think it's a good idea, completion time of `linux-amd64-cpu-opt` is much higher, it's going to delay everything behind, we should re-use `linux-amd64-ctc-opt` to build the new python wheel instead

Comment (reuben) [reply taskcluster/test-training_upstream-linux-amd64-py27mu-opt.yml]: The problem is that it's not a new Python wheel, it's several Python wheels, one for each architecture/Python version, so I tried to leverage the existing infrastructure for CPU builds.

**Comment by reuben** [comment]
Initially the compiler error mentioned -fPIC, but compiling with that flag made no difference, it still fails with relocation issues in libm.a.

**Comment by reuben** [comment]
I think the problem is that definitions.mk doesn't cover all that's needed to get a build working on ARM, as it's missing some of the stuff that was added in the tensorflow CROSSTOOL file like pointing to a different linker.

File: taskcluster/test-training_upstream-linux-amd64-py27mu-opt.yml (hunk 1 above)
Comment (lissyx) [reply taskcluster/test-training_upstream-linux-amd64-py27mu-opt.yml]: Sure, but it's still better IMHO to do that outside, we can easily hack around `do_deepspeech_python_build`

**Comment by lissyx** [comment]
> I think the problem is that definitions.mk doesn't cover all that's needed to get a build working on ARM, as it's missing some of the stuff that was added in the tensorflow CROSSTOOL file like pointing to a different linker.

It's more simply just unable to find the proper `libm` because of absolute paths in ld scripts. Quick hack that works for me:
```
diff --git a/native_client/ctcdecode/Makefile b/native_client/ctcdecode/Makefile
index 14cc591..39c5e63 100644
--- a/native_client/ctcdecode/Makefile
+++ b/native_client/ctcdecode/Makefile
@@ -4,6 +4,8 @@ include ../definitions.mk

 NUM_PROCESSES ?= 1

+LDFLAGS_NEEDED += $(RASPBIAN)/lib/aarch64-linux-gnu/libm-2.24.so
+
 all: bindings

 clean:
@@ -16,4 +18,4 @@ bindings:
        AS=$(AS) CC=$(CC) CXX=$(CXX) LD=$(LD) CFLAGS="$(CFLAGS) $(CXXFLAGS)" LDFLAGS="$(LDFLAGS_NEEDED)" $(PYTHON_PATH) $(NUMPY_INCLUDE) python ./setup.py build_ext --num_processes $(NUM_PROCESSES) $(PYTHON_PLATFORM_NAME) $(SETUP_FLAGS)
        find temp_build -type f -name "*.o" -delete
        AS=$(AS) CC=$(CC) CXX=$(CXX) LD=$(LD) CFLAGS="$(CFLAGS) $(CXXFLAGS)" LDFLAGS="$(LDFLAGS_NEEDED)" $(PYTHON_PATH) $(NUMPY_INCLUDE) python ./setup.py bdist_wheel --num_processes $(NUM_PROCESSES) $(PYTHON_PLATFORM_NAME) $(SETUP_FLAGS)
-       rm -rf temp_build
\ No newline at end of file
+       rm -rf temp_build
diff --git a/native_client/ctcdecode/setup.py b/native_client/ctcdecode/setup.py
index e6b3855..ec31f9b 100644
--- a/native_client/ctcdecode/setup.py
+++ b/native_client/ctcdecode/setup.py
@@ -89,10 +89,6 @@ LIBS = []
 ARGS = ['-O3', '-DNDEBUG', '-DKENLM_MAX_ORDER=6', '-std=c++11',
         '-Wno-unused-local-typedef', '-Wno-sign-compare']

-if os.environ['TARGET'] == 'rpi3-armv8':
-    ARGS.append('-static-libstdc++')
-
-
 decoder_module = Extension(
     name='ds_ctcdecoder._swigwrapper',
     sources=['swigwrapper.i'] + FILES + glob.glob('*.cpp'),
```

**Comment by reuben** [comment]
We have liftoff! Thanks @lissyx!

File: native_client/ctcdecode/Makefile [hunk 2]
```diff
@@ -4,6 +4,11 @@ include ../definitions.mk
 
 NUM_PROCESSES ?= 1
 
+# ARM64 can't find the proper libm.so without this
+ifeq ($(TARGET),rpi3-armv8)
+LDFLAGS_NEEDED += $(RASPBIAN)/lib/aarch64-linux-gnu/libm-2.24.so
```
Comment (lissyx) [inline native_client/ctcdecode/Makefile]: Even if it's not very nice, we should avoid adding more unniceness with hard-coding the version. `libm.so.6` should be a good stable symlink

File: native_client/ctcdecode/scorer.cpp [hunk 3]
```diff
@@ -43,6 +43,16 @@ Scorer::Scorer(double alpha,
...
 }
 
+Scorer::Scorer(double alpha,
+               double beta,
+               const std::string& lm_path,
+               const std::string& trie_path,
+               const char* alphabet_config_path)
+  : Scorer(alpha, beta, lm_path, trie_path, Alphabet(alphabet_config_path))
+{
+
```
Comment (lissyx) [inline native_client/ctcdecode/scorer.cpp]: nit: no empty line?

File: native_client/ctcdecode/scorer.cpp [hunk 4]
```diff
@@ -114,7 +124,7 @@ void Scorer::save_dictionary(const std::string& path) {
 }
 
 double Scorer::get_log_cond_prob(const std::vector<std::string>& words) {
-  double cond_prob;
+  double cond_prob = OOV_SCORE;
```
Comment (lissyx) [inline native_client/ctcdecode/scorer.cpp]: I dont see this `OOV_SCORE`, where is it coming from?

File: taskcluster/decoder-build.sh [hunk 5]
```diff
@@ -4,10 +4,7 @@ set -xe
 
 source $(dirname "$0")/../tc-tests-utils.sh
 
-source ${DS_ROOT_TASK}/DeepSpeech/tf/tc-vars.sh
+export OS=$(uname)
```
Comment (lissyx) [inline taskcluster/decoder-build.sh]: There's already this in `tc-tests-utils.sh`

Comment (reuben) [reply taskcluster/decoder-build.sh]: Maybe it's the lack of `export` there, but without this the files downstream were picking up the `export OS` in tf-vars.sh, not `tc-tests-utils.sh`. In any case, I'll add the `export` in `tc-tests-utils.sh` instead.

File: util/coordinator.py [hunk 6]
```diff
@@ -0,0 +1,706 @@
...
+                log_error('There is no running epoch of ID %d for job with ID %d. This should never happen.' % (job.epoch_id, job.id))
+            return self.get_job(job.worker)
+
+        # We are a remote worker and have to hand over to the chief worker by HTTP
+        result = self._talk_to_chief('', data=pickle.dumps(job))
+        if result:
+            result = pickle.loads(result)
+        return result
+
+class GlobalConfig:
```
Comment (tilmankamp) [inline util/coordinator.py]: This should not live here in the coordinator. You could also somehow join it with the flags module.

File: util/coordinator.py [hunk 7]
```diff
@@ -0,0 +1,706 @@
...
+    def __getattr__(self, name):
+        if not GlobalConfig._config:
+            raise RuntimeError("Global configuration not yet initialized.")
+        if not hasattr(GlobalConfig._config, name):
+            raise RuntimeError("Configuration option {} not found in config.".format(name))
+        return GlobalConfig._config[name]
+
+C = GlobalConfig()
+
+def initialize_globals():
```
Comment (tilmankamp) [inline util/coordinator.py]: This should not live here in the coordinator.

File: util/coordinator.py [hunk 8]
```diff
@@ -0,0 +1,706 @@
+from __future__ import absolute_import, division, print_function
```
Comment (tilmankamp) [inline util/coordinator.py]: It's good to have this in an extra file, as this should go away soon. We should soon move to an in-graph replica scheme for training.

File: DeepSpeech.py [hunk 9]
```diff
@@ -8,356 +8,24 @@
 log_level_index = sys.argv.index('--log_level') + 1 if '--log_level' in sys.argv else 0
 os.environ['TF_CPP_MIN_LOG_LEVEL'] = sys.argv[log_level_index] if log_level_index > 0 and log_level_index < len(sys.argv) else '3'
 
-import datetime
```
Comment (tilmankamp) [inline DeepSpeech.py]: Modularizing this file is a great thing. But why not doing it in a separate PR? I never got a similar refactoring merged due to an overloaded PR.

File: DeepSpeech.py [hunk 10]
```diff
@@ -1716,7 +586,7 @@ def update_progressbar(set_name):
 
                         log_debug('Starting batch...')
                         # Compute the batch
-                        _, current_step, batch_loss, batch_report, step_summary = session.run([train_op, global_step, loss, report_params, step_summaries_op], **extra_params)
```
Comment (tilmankamp) [inline DeepSpeech.py]: Does this mean we don't compute training and validation WER reports anymore?

File: taskcluster/host-build.sh [hunk 11]
```diff
@@ -21,6 +21,7 @@ do_deepspeech_binary_build
 
 do_deepspeech_python_build
 
+do_deepspeech_decoder_build
```
Comment (lissyx) [inline taskcluster/host-build.sh]: Why do we need this one ? In the past, there was a `decoder-build.sh` call to build the `libctc_decoder_with_kenlm.so` for packaging it, but this one does not get packaged, right?

File: DeepSpeech.py (hunk 9 above)
Comment (reuben) [reply DeepSpeech.py]: That's why I split this into commits, so you can look at each commit individually. I can also make separate PRs for the commits, but that will require making each PR buildable/passing tests, which will be a lot more work.

File: .install [hunk 12]
```diff
@@ -7,4 +7,10 @@ pip install tensorflow-gpu==1.12.0rc2
 
 python3 util/taskcluster.py --arch gpu --target ../tmp/native_client
 
+# Install ds_ctcdecoder package from TaskCluster
+VERSION=$(python -c 'import pkg_resources; print(pkg_resources.safe_version(open("VERSION").read()))')
+PYVER=$(python -c 'import sys; print("cp{0}{1}-cp{0}{1}m".format(sys.version_info.major, sys.version_info.minor))')
+python3 util/taskcluster.py --arch cpu --target ../tmp --artifact "ds_ctcdecoder-${VERSION}-${PYVER}-manylinux1_x86_64.whl"
```
Comment (lissyx) [inline .install]: this will install from `master` then ?

File: DeepSpeech.py (hunk 9 above)
Comment (tilmankamp) [reply DeepSpeech.py]: I know - that's also the reason why I don't insist in doing it. :-) Doing this after the fact is much harder and a big waste of time.

File: taskcluster/host-build.sh (hunk 11 above)
Comment (reuben) [reply taskcluster/host-build.sh]: Good point, I don't think this is needed.

File: util/coordinator.py (hunk 6 above)
Comment (reuben) [reply util/coordinator.py]: There's a circular dependency between the globals and the coordinator (the COORD global) and IMO breaking it by separating just that global initialization out of `initialize_globals` would be worse than having these live in the same place. What do you think?

Comment (tilmankamp) [reply util/coordinator.py]: I'd initalize the `COORD` global variable in `DeepSpeech.py` (moving the block with comment `# Initializing and starting the training coordinator` back there). Then there should be a clean separation of concerns again.

**Comment by reuben** [comment]
In 38b5447 I've updated the docs and added some convenience code in `util/taskcluster.py` to get the URL of the `ds_ctcdecoder` package for easy installation.

File: evaluate.py [hunk 13]
```diff
@@ -15,8 +15,10 @@
 from attrdict import AttrDict
 from collections import namedtuple
 from ds_ctcdecoder import ctc_beam_search_decoder_batch, Scorer
-from DeepSpeech import initialize_globals, create_flags, log_debug, log_info, log_warn, log_error, create_inference_graph
-from multiprocessing import Pool
+from util.flags import create_flags
+from util.coordinator import C, initialize_globals
```
Comment (kdavis-mozilla) [inline evaluate.py]: Again the naming of `C` seems too generic.

File: evaluate.py [hunk 14]
```diff
@@ -221,7 +206,38 @@ def create_windows(features):
...
+        FLAGS.test_files.split(','),
+        FLAGS.test_batch_size,
+        alphabet=alphabet,
+        numcep=N_FEATURES,
+        numcontext=N_CONTEXT,
+        hdf5_cache_path=FLAGS.hdf5_test_set).sort_values(
+        by="features_len",
+        ascending=False)
+
+    from DeepSpeech import create_inference_graph
```
Comment (kdavis-mozilla) [inline evaluate.py]: Why import here and not at file's top?

File: native_client/ctcdecode/scorer.cpp (hunk 4 above)
Comment (kdavis-mozilla) [reply native_client/ctcdecode/scorer.cpp]: It's in `scorer.h`

File: util/taskcluster.py [hunk 15]
```diff
@@ -74,9 +78,55 @@ def maybe_download_tc_bin(**kwargs):
...
+        plat = platform.system().lower()
+        arch = platform.machine()
+
+        if plat == 'linux' and arch == 'x86_64':
+            plat = 'manylinux1'
+
+        if plat == 'darwin':
+            plat = 'macosx_10_10'
+
+        from pkg_resources import parse_version
```
Comment (kdavis-mozilla) [inline util/taskcluster.py]: Just wondering is there is a reason to use local imports instead of globals.

File: util/coordinator.py [hunk 16]
```diff
@@ -186,7 +184,7 @@ def done(self):
 
                 # if the job was for validation dataset then append it to the COORD's _loss for early stop verification
                 if (FLAGS.early_stop is True) and (self.set_name == 'dev'):
-                    COORD._dev_losses.append(self.loss)
+                    self.coord_dev_losses.append(self.loss)
```
Comment (kdavis-mozilla) [inline util/coordinator.py]: This should be
```python
self.coord._dev_losses.append(self.loss)
```

File: util/coordinator.py [hunk 17]
```diff
@@ -415,15 +417,15 @@ def start(self):
         if self.is_chief:
             log_debug('Starting coordinator...')
             self._thread = Thread(target=self._httpd.serve_forever)
-            self._thread.daemon = True
+            # self._thread.daemon = True
```
Comment (kdavis-mozilla) [inline util/coordinator.py]: Out of curiosity, why did this change?

File: evaluate.py (hunk 14 above)
Comment (reuben) [reply evaluate.py]: That would make it a circular dependency as DeepSpeech.py imports evaluate.py.

File: util/coordinator.py (hunk 17 above)
Comment (reuben) [reply util/coordinator.py]: Leftover from debugging, will remove this.

File: DeepSpeech.py [hunk 18]
```diff
@@ -8,356 +8,24 @@
...
+from ds_ctcdecoder import ctc_beam_search_decoder, Scorer
+from six.moves import zip, range
 from tensorflow.contrib.lite.python import tflite_convert
-from threading import Thread, Lock
+from tensorflow.python.tools import freeze_graph
 from util.audio import audiofile_to_input_vector
 from util.feeding import DataSet, ModelFeeder
+from util.logging import *
+from util.flags import create_flags, FLAGS
+from util.coordinator import C, initialize_globals
```
[the remark this answers is not in the document]
Comment (reuben) [reply DeepSpeech.py]: The reason for the short name is to make it easier to transition our generous use of globals to more reasonable solutions. A long name would screw up code style across the project. But if you feel strongly about this I'll rename it to Config or something like that.

File: DeepSpeech.py [hunk 19]
```diff
@@ -1948,8 +824,10 @@ def __init__(self):
             log_error(str(e))
 
 def do_single_file_inference(input_file_path):
-    with tf.Session(config=session_config) as session:
-        inputs, outputs, _ = create_inference_graph(batch_size=1, use_new_decoder=True)
+    import numpy as np
+
+    with tf.Session(config=C.session_config) as session:
+        inputs, outputs, _ = create_inference_graph(batch_size=1, n_steps=-1, use_new_decoder=True)
```
[the remark this answers is not in the document]
Comment (reuben) [reply DeepSpeech.py]: This was already unused before this PR. Removed.

**Comment by reuben** [comment]
Everything should be addressed now.

**Comment by kdavis-mozilla** [comment]
LGTM

File: native_client/ctcdecode/Makefile [hunk 20]
```diff
@@ -6,7 +6,7 @@ NUM_PROCESSES ?= 1
 
 # ARM64 can't find the proper libm.so without this
 ifeq ($(TARGET),rpi3-armv8)
-LDFLAGS_NEEDED += $(RASPBIAN)/lib/aarch64-linux-gnu/libm-2.24.so
+LDFLAGS_NEEDED += $(RASPBIAN)/lib/aarch64-linux-gnu/libm.so.6
```
Comment (lissyx) [inline native_client/ctcdecode/Makefile]: :+1:

File: util/taskcluster.py [hunk 21]
```diff
@@ -59,13 +60,16 @@ def maybe_download_tc_bin(**kwargs):
...
 if __name__ == '__main__':
     import argparse
 
     parser = argparse.ArgumentParser(description='Tooling to ease downloading of components from TaskCluster.')
-    parser.add_argument('--target', required=True,
+    parser.add_argument('--target', required=False,
                         help='Where to put the native client binary files')
-    parser.add_argument('--arch', required=False, default='cpu',
+    parser.add_argument('--arch', required=False,
                         help='Which architecture to download binaries for. "arm" for ARM 7 (32-bit), "gpu" for CUDA enabled x86_64 binaries, "cpu" for CPU-only x86_64 binaries, "osx" for CPU-only x86_64 OSX binaries. Optional ("cpu" by default)')
```
Comment (lissyx) [inline util/taskcluster.py]: nit: also mention arm64 here, since we its default value

**Comment by reuben** [comment]
Yes.

-- reuben

**Comment by JRMeyer** [comment]
@reuben I'm loading in an older checkpoint (v0.3.0), and I get the following error:

```
KeyError: 'CTCBeamSearchDecoderWithLM'
````

After some searching, I think this is related to a missing `libctc_decoder_with_kenlm.so` file (c.f. #1306, #950)

 The following is the script I'm running to load the checkpoint:


```
import tensorflow as tf
from tensorflow.python.tools import freeze_graph
from ds_ctcdecoder import ctc_beam_search_decoder, Scorer

sess = tf.Session()
sess.run(tf.global_variables_initializer())
 
tf.contrib.rnn

tf.load_op_library('../tmp/native_client/libdeepspeech.so')

new_saver = tf.train.import_meta_graph('../v030/ckpt/model.v0.2.0.meta')
new_saver.restore(sess, '../v030/ckpt/model.v0.2.0')
```

Is this error a by-product of this PR? Are the older checkpoints on the release page not forwards compatible after this PR?

-josh

**Comment by JRMeyer** [comment]
Strange, but good news.

**Pull request merged**